alembic/env.py

```
@@ -20,8 +20,13 @@
     fileConfig(config.config_file_name)
 
 # Get database URL from environment variable
-# This ensures no hardcoded credentials in the codebase
+# In CI, fail closed: migrations must use MIGRATION_DATABASE_URL explicitly.
 migration_database_url = os.environ.get("MIGRATION_DATABASE_URL")
+ci_mode = os.environ.get("CI", "").lower() == "true"
+if ci_mode and not migration_database_url:
+    raise ValueError(
+        "MIGRATION_DATABASE_URL is required in CI for strict migration/runtime identity separation."
+    )
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)